// vite-plugin-rsw/dist/index.js
"use strict";var I=Object.create;var l=Object.defineProperty;var O=Object.getOwnPropertyDescriptor;var C=Object.getOwnPropertyNames;var H=Object.getPrototypeOf,P=Object.prototype.hasOwnProperty;var $=(e,r)=>{for(var o in r)l(e,o,{get:r[o],enumerable:!0})},m=(e,r,o,s)=>{if(r&&typeof r=="object"||typeof r=="function")for(let t of C(r))!P.call(e,t)&&t!==o&&l(e,t,{get:()=>r[t],enumerable:!(s=O(r,t))||s.enumerable});return e};var w=(e,r,o)=>(o=e!=null?I(H(e)):{},m(r||!e||!e.__esModule?l(o,"default",{value:e,enumerable:!0}):o,e)),z=e=>m(l({},"__esModule",{value:!0}),e);var G={};$(G,{ViteRsw:()=>T,default:()=>M});module.exports=z(G);var i=w(require("fs")),c=w(require("path")),y=w(require("readline")),n={status:"",name:"",build:"",error:"",path:""},g=c.default.resolve(process.cwd(),".rsw","rsw.info"),L=c.default.resolve(process.cwd(),".rsw","rsw.err"),x=c.default.resolve(process.cwd(),".rsw","rsw.crates"),v=e=>e.split(" :~> ")||[],b=()=>{let e=[],r=[];return i.default.existsSync(x)&&i.default.readFileSync(x,{encoding:"utf8"}).split(`
`).forEach(s=>{let t=v(s);t.length>0&&(t!=null&&t[0]&&e.push(t==null?void 0:t[0]),t!=null&&t[1]&&r.push(t==null?void 0:t[1]))}),{crates:e,cratesPath:r}},R=(e,r)=>{i.default.watchFile(g,{bigint:!1,interval:300,persistent:!0},()=>{if(!i.default.existsSync(g))return;let o=y.default.createInterface({input:i.default.createReadStream(g),crlfDelay:1/0});o.on("line",s=>{let[,t]=v(s);switch(!0){case/\[RSW::OK\]/.test(s):n.status="ok";break;case/\[RSW::ERR\]/.test(s):n.status="err";break;case/\[RSW::NAME\]/.test(s):n.name=t;break;case/\[RSW::PATH\]/.test(s):n.path=t;break;case/\[RSW::BUILD\]/.test(s):n.build=t;break}}),o.on("close",()=>{n.status==="err"?n.error=i.default.readFileSync(L,{encoding:"utf8"}):n.error="",r&&e(n),r=!1})})};var u=class{constructor(){this.line="",this.index=-1,this.msgTagGroup=[]}init(r,o){return this.line=r,this.index=o,this.msgTagGroup[o]=r,this}setTag(r,o){new RegExp(o).test(this.line)&&(this.msgTagGroup[this.index]=this.line.replace(o,`<code class="rsw-${r}">$1</code>`))}handle(r,o){return this.setTag(r,o),this}getValue(){return this.msgTagGroup.join(`
`)}};function h(e){let r=new u;return e.split(`
`).forEach((o,s)=>{r.init(o,s).handle("line",/(^\s+-->|\s*=|[\s\d]*\|)/).handle("compiling",/(^\s+Compiling)/).handle("error",/(^error(\[\w+\])?)/).handle("warn",/(^warning)/).handle("help",/(^help)/)}),r.getValue()}var E=`
if (import.meta.hot) {
  import.meta.hot.on('rsw-error', (data) => {
    createRswErrorOverlay && createRswErrorOverlay(data);
    throw \`
\u{1F980}\${data.plugin} ~> \${data.id}

\${data.console}\`
  })
}`,k=`
const rswTemplate = \`
<style>
:host {
  position: fixed;
  z-index: 99999;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  overflow-y: scroll;
  margin: 0;
  background: rgba(0, 0, 0, 0.66);
  --monospace: 'SFMono-Regular', Consolas,
              'Liberation Mono', Menlo, Courier, monospace;
  --red: #ff5555;
  --green: #26cb7c;
  --yellow: #e2aa53;
  --purple: #cfa4ff;
  --cyan: #2dd9da;
  --dim: #c9c9c9;
  --blue: #3884ff;
}
.window {
  font-family: var(--monospace);
  line-height: 1.5;
  width: 800px;
  color: #d8d8d8;
  margin: 30px auto;
  padding: 25px 40px;
  position: relative;
  background: #181818;
  border-radius: 6px 6px 8px 8px;
  box-shadow: 0 19px 38px rgba(0,0,0,0.30), 0 15px 12px rgba(0,0,0,0.22);
  overflow: hidden;
  border-top: 8px solid var(--red);
}
pre {
  font-family: var(--monospace);
  font-size: 14px;
  margin-top: 0;
  margin-bottom: 1em;
  overflow-x: scroll;
  scrollbar-width: none;
}
pre::-webkit-scrollbar {
  display: none;
}
.message {
  line-height: 1.3;
  color: #6d7878;
  font-size: 14px;
}
.plugin {
  color: var(--purple);
  font-weight: bold;
}
.file {
  color: var(--green);
  margin: 8px 0;
  white-space: pre-wrap;
  word-break: break-all;
  font-size: 14px;
  font-weight: bold;
  text-decoration: underline;
  cursor: pointer;
}
.tip {
  font-size: 13px;
  color: #999;
  border-top: 1px dotted #999;
  padding-top: 13px;
}
code {
  font-size: 13px;
  font-family: var(--monospace);
  font-weight: bold;
}
.rsw-line {
  color: var(--blue);
}
.rsw-compiling {
  color: var(--green);
}
.rsw-error {
  color: var(--red);
}
.rsw-warn {
  color: var(--yellow);
}
.rsw-help {
  color: var(--cyan);
}
.file-link {
  text-decoration: underline;
  cursor: pointer;
}
</style>
<div class="window">
  <span class="plugin"></span>
  <pre class="file"></pre>
  <pre class="message"></pre>
  <div class="tip">
    [rsw::error] This error occurred during the build time, click outside or fix the code to dismiss.
  </div>
</div>
\`;
class RswErrorOverlay extends HTMLElement {
  constructor(payload) {
    super()
    this.root = this.attachShadow({ mode: 'open' });
    this.root.innerHTML = rswTemplate;
    this.text('.message', payload.message.trim());
    this.text('.plugin', payload.plugin.trim());
    this.text('.file', payload.id.trim());
    this.root.querySelector('.window').addEventListener('click', (e) => {
      e.stopPropagation();
    });
    this.addEventListener('click', () => {
      this.close();
    });
  }
  text(selector, text) {
    const el = this.root.querySelector(selector);
    if (el) el.innerHTML = text;
  }
  close() {
    if (this.parentNode) this.parentNode.removeChild(this);
  }
}
const overlayRswId = 'vite-rsw-error-overlay';
if (!customElements.get(overlayRswId)) {
  customElements.define(overlayRswId, RswErrorOverlay);
}
function createRswErrorOverlay(err) {
  clearRswErrorOverlay();
  document.body.appendChild(new RswErrorOverlay(err));
}
function clearRswErrorOverlay() {
  document
    .querySelectorAll(overlayRswId)
    .forEach((n) => n.close());
}
window.createRswErrorOverlay = createRswErrorOverlay;`;var S=e=>e.replace(/[\\/]+/g,"/").replace(/^([a-zA-Z]+:|\.\/)/,"");function T(){let e,{cratesPath:r}=b(),o=r.map(s=>`${s}/.*.js`).join("|");return{name:"vite-plugin-rsw",enforce:"pre",apply:"serve",configResolved(s){e=s},handleHotUpdate({file:s,server:t}){let p=!0,d=S(s);/\/target(\/[\d\w-_]+)?\/debug\//.test(d)||/\.rsw\//.test(d)||/\.rs$/.test(d)&&R(a=>{a.status==="ok"&&t.ws.send({type:"full-reload",path:"*"}),a.status==="err"&&t.ws.send({type:"custom",event:"rsw-error",data:{plugin:"[vite::rsw]",message:h(a.error),id:a.path,console:a.error}})},p)},transform(s,t){let p=S(t);return new RegExp(`${o}`).test(p)?s+E:s},transformIndexHtml(s){return(e==null?void 0:e.mode)==="development"?[{tag:"script",attrs:{type:"module"},children:k}]:s}}}var M=T;0&&(module.exports={ViteRsw});
